File: Data Structures & Algorithms/subsets/submission-0.py
class Solution:    
    def subsets(self, nums: List[int]) -> List[List[int]]:
        """
            1 2 3
        """
        ans = []
        def backtrack(curr: int, acc: List[int]):
            if curr >= len(nums):
                ans.append(acc[:])
                return
            

            acc.append(nums[curr])
            backtrack(curr+1,acc)
            acc.pop()

            backtrack(curr+1,acc)

        backtrack(0,[])
        return ans

File: Data Structures & Algorithms/subsets-ii/submission-0.cpp
class Solution {
public:
    vector<vector<int>>result;
    void dfs(vector<int>& nums,int curr,int n,vector<int>path){
        if(curr >= n){
            if(count(result.begin(),result.end(),path) == 0) result.push_back(path);
            return;
        }

        path.push_back(nums[curr]);
        dfs(nums,curr+1,n,path);
        path.pop_back();

        dfs(nums,curr+1,n,path);

    }

    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        //all subsets backtracking and recursion
        result.clear();

        sort(nums.begin(),nums.end());
        int n = nums.size();
        dfs(nums,0,n,{});
        return result;
    }
};
